compact_str/src/lib.rs

```diff
@@ -1170,12 +1173,98 @@ impl CompactString {
+        from_x_bytes: impl Fn([u8; 2]) -> u16,
+    ) -> Result<Self, Utf16Error> {
+        match unsafe { v.align_to::<u16>() } {
+            (&[], _, &[_]) | (&[_], _, &[]) => {
+                // Input had an odd number of bytes.
+                Err(Utf16Error(()))
+            }
+            (&[], v, _) => {
+                // Input is correcty aligned.
+                std::char::decode_utf16(v.iter().copied().map(from_x))
+                    .collect::<Result<CompactString, _>>()
+                    .map_err(|_| Utf16Error(()))
+            }
+            _ => {
+                // Input's alignment is off.
+                // SAFETY: we can always reinterpret a `[u8; 2*N]` slice as `[[u8; 2]; N]`
+                let v = unsafe { slice::from_raw_parts(v.as_ptr().cast(), v.len() / 2) };
+                std::char::decode_utf16(v.iter().copied().map(from_x_bytes))
+                    .collect::<Result<CompactString, _>>()
+                    .map_err(|_| Utf16Error(()))
+            }
+        }
+    }
+
+    /// Decode a slice of bytes as UTF-16 encoded string, in little endian.
+    ///
+    /// # Errors
+    ///
+    /// If the slice has an odd number of bytes, or if it did not contain valid UTF-16 characters,
+    /// a [`Utf16Error`] is returned.
+    ///
+    /// # Examples
+    ///
+    /// ```
+    /// # use compact_str::CompactString;
+    /// const DANCING_MEN: &[u8] = b"\x3d\xd8\x6f\xdc\x0d\x20\x42\x26\x0f\xfe";
+    /// let dancing_men = CompactString::from_utf16le(DANCING_MEN).unwrap();
+    /// assert_eq!(dancing_men, "👯‍♂️");
+    /// ```
+    pub fn from_utf16le(v: &[u8]) -> Result<Self, Utf16Error> {
+        Self::from_utf16x(v, u16::from_le, u16::from_le_bytes)
+    }
+
+    /// Decode a slice of bytes as UTF-16 encoded string, in big endian.
+    ///
+    /// # Errors
+    ///
+    /// If the slice has an odd number of bytes, or if it did not contain valid UTF-16 characters,
+    /// a [`Utf16Error`] is returned.
+    ///
+    /// # Examples
+    ///
+    /// ```
+    /// # use compact_str::CompactString;
+    /// const DANCING_WOMEN: &[u8] = b"\xd8\x3d\xdc\x6f\x20\x0d\x26\x40\xfe\x0f";
+    /// let dancing_women = CompactString::from_utf16be(DANCING_WOMEN).unwrap();
+    /// assert_eq!(dancing_women, "👯‍♀️");
+    /// ```
+    pub fn from_utf16be(v: &[u8]) -> Result<Self, Utf16Error> {
+        Self::from_utf16x(v, u16::from_be, u16::from_be_bytes)
+    }
+
+    /// Decode a slice of bytes as UTF-16 encoded string, in native endian.
+    ///
+    /// # Errors
+    ///
+    /// If the slice has an odd number of bytes, or if it did not contain valid UTF-16 characters,
+    /// a [`Utf16Error`] is returned.
+    ///
+    /// # Examples
+    ///
+    /// Please see [`CompactString::from_utf16le()`] and [`CompactString::from_utf16be()`]
+    /// for examples.
+    #[inline(always)]
+    #[cfg(any(target_endian = "little", target_endian = "big"))]
+    pub fn from_utf16ne(v: &[u8]) -> Result<Self, Utf16Error> {
```

Comment:
Thinking about this a bit further, do you think we actually need a native endian version of this function?

I haven't worked with UTF-16 that much, but my impression is if someone needs to be specific about the endianness that's because some protocol specifies UTF-16LE or UTF-16BE. I want to prevent any overlap between `from_utf16(buf: &[u16])` and `from_utf16ne(buf: &[u8])` to prevent confusion about which method should be used